### PR TITLE
upgrade CachedUPAKLoader to upak V2

### PR DESCRIPTION
**Pull request by oconnor663**
r? @maxtaco 

@patrickxb, do you want to review UPAK loader code also?

### CODE REVIEW

File: go/libkb/rpc_exim.go [hunk 1]
```diff
@@ -915,7 +915,7 @@ func publicKeyV2BaseFromComputedKeyInfo(info ComputedKeyInfo) (base keybase1.Pub
 	}
 	if info.DelegatedAt != nil {
 		base.Provisioning = keybase1.SignatureMetadata{
-			Time: keybase1.Time(info.DelegatedAt.Unix),
+			Time: keybase1.TimeFromSeconds(info.DelegatedAt.Unix),
```
Comment (oconnor663) [inline go/libkb/rpc_exim.go]: Converting to UpakV2 revealed a couple bugs in my converter function. The first is that I wasn't converting these two times properly. The second was that I was assembling an unsorted tracks list, which broke a `sort.Search` that happens later (that took forever to figure out :p).

File: go/libkb/sig_chain_test.go [hunk 2]
```diff
@@ -342,15 +340,11 @@ func storeAndLoad(t *testing.T, tc TestContext, chain *SigChain) {
 		t.Fatal(err)
 	}
 
-	// Loading a chain from cache won't load all the links, if there's a
-	// sigchain reset in the middle. So we don't want to check the entire
-	// loaded chain for equality with what we saved. But we can make sure it's
-	// loaded all the links of the *current subchain*. HOWEVER: Loading
```
Comment (oconnor663) [inline go/libkb/sig_chain_test.go]: The first half of this comment is no longer true, and we can now test the whole chain.

File: go/libkb/upak.go [hunk 3]
```diff
@@ -23,35 +23,33 @@ func BaseProofSet(u *keybase1.UserPlusAllKeys) *ProofSet {
...
-			return true, nil, true
+func checkKIDKeybase(u *keybase1.UserPlusKeysV2AllIncarnations, kid keybase1.KID) (found bool, revokedAt *keybase1.KeybaseTime, deleted bool) {
+	for _, key := range u.Current.DeviceKeys {
+		if key.Base.Kid.Equal(kid) {
+			found = true
+			if key.Base.Revocation != nil {
+				// This is the inverse of the marshalling we do in rpc_exim.go.
+				revokedAt = &keybase1.KeybaseTime{
+					Unix:  key.Base.Revocation.Time,
+					Chain: int(key.Base.Revocation.PrevMerkleRootSigned.Seqno),
```
Comment (oconnor663) [inline go/libkb/upak.go]: We don't cast the time here (that would usually be a mistake), but we do cast the Chain seqno. Hopefully that's ok?

File: go/libkb/upak_loader.go [hunk 4]
```diff
@@ -188,19 +193,20 @@ func (u *CachedUPAKLoader) putUPAKToCache(ctx context.Context, obj *keybase1.Use
...
 	return err
 }
 
 func (u *CachedUPAKLoader) PutUserToCache(ctx context.Context, user *User) error {
 
 	lock := u.locktab.AcquireOnName(ctx, u.G(), user.GetUID().String())
 	defer lock.Release(ctx)
-	upak := user.ExportToUserPlusAllKeys(keybase1.Time(0))
-	upak.Base.Uvv.CachedAt = keybase1.ToTime(u.G().Clock().Now())
+	upak := user.ExportToUPKV2AllIncarnations(keybase1.Time(0))
```
Comment (oconnor663) [inline go/libkb/upak_loader.go]: Is `keybase1.Time(0)` really correct here?

File: go/libkb/db.go [hunk 5]
```diff
@@ -166,7 +166,8 @@ const (
 	DBUser                    = 0x00
 	DBSig                     = 0x0f
 	DBTeamChain               = 0x10
-	DBUserPlusAllKeys         = 0x19
+	DBUserPlusAllKeysV1       = 0x19
+	DBUserPlusKeysAIV2        = 0x1a
```
Comment (patrickxb) [inline go/libkb/db.go]: `AI` is AllIncarnations?

**Comment by oconnor663** [comment]
Yep

**Comment by patrickxb** [comment]
LGTM, but defer to @maxtaco overall

**Comment by oconnor663** [comment]
That test failure is real, looking at it.

File: go/libkb/upak.go [hunk 6]
```diff
@@ -23,35 +23,45 @@ func BaseProofSet(u *keybase1.UserPlusAllKeys) *ProofSet {
...
-	for _, key := range u.Base.DeviceKeys {
-		if key.KID.Equal(kid) {
-			return true, nil, false
+func checkKIDKeybase(u *keybase1.UserPlusKeysV2AllIncarnations, kid keybase1.KID) (found bool, revokedAt *keybase1.KeybaseTime, deleted bool) {
+	exportRevokedAt := func(key keybase1.PublicKeyV2NaCl) *keybase1.KeybaseTime {
+		if key.Base.Revocation != nil {
+			// This is the inverse of the marshalling we do in rpc_exim.go.
+			return &keybase1.KeybaseTime{
+				Unix:  key.Base.Revocation.Time,
+				Chain: int(key.Base.Revocation.PrevMerkleRootSigned.Seqno),
```
Comment (oconnor663) [inline go/libkb/upak.go]: We don't cast the time here (that would usually be a mistake), but we do cast the Chain seqno. Hopefully that's ok?

**Comment by oconnor663** [comment]
Test failure was that I was no longer returning `(found, deleted) = (true, true)` for deleted keys in `CheckKID`. Should be fixed now.

**Comment by oconnor663** [comment]
Waiting for final review from @maxtaco

File: go/libkb/db.go (hunk 5 above)
Comment (maxtaco) [inline go/libkb/db.go]: Can't take `0x1a`, that already means `billing_plan`.  If it doesn't exist in `id.iced`, I'd use `0xe4`

Comment (maxtaco) [inline go/libkb/db.go]: Can you store `UPAKVersioned` here btw to prevent this problem in the future?

File: go/libkb/rpc_exim.go [hunk 7]
```diff
@@ -1201,16 +1201,8 @@ func (u *User) ExportToUPKV2AllIncarnations(idTime keybase1.Time) keybase1.UserP
...
-	for _, track := range u.IDTable().GetTrackList() {
-		remoteTracks = append(remoteTracks, keybase1.RemoteTrack{
-			Username: string(track.whomUsername),
-			Uid:      track.whomUID,
-			LinkID:   keybase1.LinkID(hex.EncodeToString(track.LinkID())),
-		})
-	}
-	current.RemoteTracks = remoteTracks
+	// NOTE: This list *must* be in sorted order. If we ever write V3, be careful to keep it sorted!
+	current.RemoteTracks = u.ExportRemoteTracks()
```
Comment (maxtaco) [inline go/libkb/rpc_exim.go]: Do you want to change this field to a `map` to avoid this problem?

File: go/libkb/sig_chain.go [hunk 8]
```diff
@@ -646,11 +645,10 @@ func (sc *SigChain) verifySigsAndComputeKeysCurrent(ctx context.Context, eldest
 		return cached, 0, err
 	}
 
-	if sc.allKeys || sc.loadedFromLinkOne {
-		if first := sc.getFirstSeqno(); first > keybase1.Seqno(1) {
-			err = ChainLinkWrongSeqnoError{fmt.Sprintf("Wanted a chain from seqno=1, but got seqno=%d", first)}
-			return cached, 0, err
-		}
+	// AllKeys mode is now the default.
+	if first := sc.getFirstSeqno(); first > keybase1.Seqno(1) {
```
Comment (maxtaco) [inline go/libkb/sig_chain.go]: will this fail for previously cached users?

File: go/libkb/upak.go (hunk 6 above)
Comment (maxtaco) [inline go/libkb/upak.go]: can you change `KeybaseTime` to use a `keybase1.Seqno`?

Comment (maxtaco) [inline go/libkb/upak.go]: ... that might be a hornet's nest...

**Comment by oconnor663** [comment]
The `Seqno` change turned out to be surprisingly easy.

File: go/libkb/sig_chain.go (hunk 8 above)
Comment (oconnor663) [inline go/libkb/sig_chain.go]: Just checked by manually deleting some old links, and I think we're fine. The LoadFromStorage step will short circuit and return `nil` if any of the links we're trying to load are missing ([here](https://github.com/keybase/client/blob/e97bab9/go/libkb/sig_chain.go#L914) and [here](https://github.com/keybase/client/blob/e97bab9/go/libkb/sig_chain.go#L935)), which will cause a new server fetch of the entire chain.

**Comment by maxtaco** [comment]
In CI above looks like `TestCacheFallbacks` failed, I've never seen that fail before.

File: go/libkb/rpc_exim.go [hunk 9]
```diff
@@ -1201,29 +1201,20 @@ func (u *User) ExportToUPKV2AllIncarnations(idTime keybase1.Time) keybase1.UserP
...
-
-	remoteTracks := []keybase1.RemoteTrack{}
-	for _, track := range u.IDTable().GetTrackList() {
-		remoteTracks = append(remoteTracks, keybase1.RemoteTrack{
-			Username: string(track.whomUsername),
-			Uid:      track.whomUID,
-			LinkID:   keybase1.LinkID(hex.EncodeToString(track.LinkID())),
-		})
+	current.RemoteTracks = make(map[keybase1.UID]keybase1.RemoteTrack)
+	if u.IDTable() != nil {
```
Comment (maxtaco) [inline go/libkb/rpc_exim.go]: when would `IDTable()` be nil?

File: go/libkb/upak_loader.go (hunk 4 above)
Comment (maxtaco) [inline go/libkb/upak_loader.go]: Remind me what it means?

File: go/libkb/rpc_exim.go (hunk 9 above)
Comment (oconnor663) [inline go/libkb/rpc_exim.go]: It's nil if the user has no keys at all.

**Comment by maxtaco** [comment]
👍 Once CI passes

**Comment by maxtaco** [comment]
Also please hand-audit the JSON stored into the DB, to make sure it's legit.

**Comment by keybase-ci-visdiff** [comment]
The commits ce79d7f7ce76...435219cc77fd introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Tracker-New_user_Scroll_1** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1497994919278-ce79d7f7ce76-435219cc77fd-linux/Tracker-New_user_Scroll_1-ce79d7f7ce76.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1497994919278-ce79d7f7ce76-435219cc77fd-linux/Tracker-New_user_Scroll_1-435219cc77fd.png)  
   Tracker-New_user_Scroll_1 comparison
</summary>

**Comment by oconnor663** [comment]
Hmm, a lot of these are passing locally. Will keep playing with it.

**Comment by keybase-ci-visdiff** [comment]
The commits 34372bc78cbd...1197568f5d04 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Markdown-emoji** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1498157585182-34372bc78cbd-1197568f5d04-linux/Markdown-emoji-34372bc78cbd.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1498157585182-34372bc78cbd-1197568f5d04-linux/Markdown-emoji-1197568f5d04.png)  
   Markdown-emoji comparison
</summary>

**Comment by oconnor663** [comment]
The time we pass to the export function is `idTime`. It looks like this gets overwritten when we export a user directly from the identify2 engine, but otherwise never gets set.

**Comment by maxtaco** [comment]
If you want to delete it, go ahead!

**Comment by oconnor663** [comment]
Example of json going into the DB:

```json
{
    "current": {
        "uid": "b8f4b6d008ab99074a7b7d20510b8419",
        "username": "pukup_e662204382",
        "eldestSeqno": 1,
        "perUserKeys": [
            {
                "gen": 1,
                "seqno": 5,
                "sigKID": "0120cc9787d643bf3f45e7b6f918690c11640e8ff4780415878fa4ff1c9b3f7318a00a",
                "encKID": "01214f916c2eda485dadf02099ac12f6fa4225d852487e8e41563d8ff6b91469f54a0a"
            }
        ],
        "deviceKeys": {
            "012035d63ec0989096b846b84eadf5905ed63ad93772a8ae0c52bfee76afccb45f0f0a": {
                "base": {
                    "kid": "012035d63ec0989096b846b84eadf5905ed63ad93772a8ae0c52bfee76afccb45f0f0a",
                    "isSibkey": true,
                    "isEldest": false,
                    "cTime": 1498245384000,
                    "eTime": 2002821384000,
                    "provisioning": {
                        "signingKID": "0120be2e5f447c415cf95096adbd5a16ef3c82b96b91b4039a892072272d4f084f040a",
                        "prevMerkleRootSigned": {
                            "seqno": 78105,
                            "hashMeta": "81c3d5be0cd573177d9983cbd45fd7bbd79a79c96cf04a9a70f1b50a524657e0"
                        },
                        "firstAppearedUnverified": 78106,
                        "time": 1498245384000
                    }
                },
                "deviceID": "5024300e92f0d0284be6bca92456d418",
                "deviceDescription": "diet helmet",
                "deviceType": "backup"
            },
            "0120be2e5f447c415cf95096adbd5a16ef3c82b96b91b4039a892072272d4f084f040a": {
                "base": {
                    "kid": "0120be2e5f447c415cf95096adbd5a16ef3c82b96b91b4039a892072272d4f084f040a",
                    "isSibkey": true,
                    "isEldest": true,
                    "cTime": 1498245384000,
                    "eTime": 2002821384000,
                    "provisioning": {
                        "signingKID": "",
                        "prevMerkleRootSigned": {
                            "seqno": 0,
                            "hashMeta": ""
                        },
                        "firstAppearedUnverified": 0,
                        "time": 0
                    }
                },
                "deviceID": "5d1a53961fc547a1037f5f49dbb5b518",
                "deviceDescription": "my device",
                "deviceType": "desktop"
            },
            "01216551fc2dd564a14388d63d3ac1cfaec7278f317e410b87c3c65eae85c917dd470a": {
                "base": {
                    "kid": "01216551fc2dd564a14388d63d3ac1cfaec7278f317e410b87c3c65eae85c917dd470a",
                    "isSibkey": false,
                    "isEldest": false,
                    "cTime": 1498245384000,
                    "eTime": 2002821384000,
                    "provisioning": {
                        "signingKID": "012035d63ec0989096b846b84eadf5905ed63ad93772a8ae0c52bfee76afccb45f0f0a",
                        "prevMerkleRootSigned": {
                            "seqno": 78105,
                            "hashMeta": "81c3d5be0cd573177d9983cbd45fd7bbd79a79c96cf04a9a70f1b50a524657e0"
                        },
                        "firstAppearedUnverified": 78106,
                        "time": 1498245384000
                    }
                },
                "parent": "012035d63ec0989096b846b84eadf5905ed63ad93772a8ae0c52bfee76afccb45f0f0a",
                "deviceID": "",
                "deviceDescription": "",
                "deviceType": ""
            },
            "0121b2914e7cdbee75cc1cbde881965f37584e89b8f09497374d3186c2b34914c3410a": {
                "base": {
                    "kid": "0121b2914e7cdbee75cc1cbde881965f37584e89b8f09497374d3186c2b34914c3410a",
                    "isSibkey": false,
                    "isEldest": false,
                    "cTime": 1498245384000,
                    "eTime": 2002821384000,
                    "provisioning": {
                        "signingKID": "0120be2e5f447c415cf95096adbd5a16ef3c82b96b91b4039a892072272d4f084f040a",
                        "prevMerkleRootSigned": {
                            "seqno": 78104,
                            "hashMeta": "d72efec0f917438a04a3a1e412e08fdc2bce5485e63ef46d22269896fbd79d4b"
                        },
                        "firstAppearedUnverified": 78105,
                        "time": 1498245384000
                    }
                },
                "parent": "0120be2e5f447c415cf95096adbd5a16ef3c82b96b91b4039a892072272d4f084f040a",
                "deviceID": "",
                "deviceDescription": "",
                "deviceType": ""
            }
        },
        "pgpKeys": {

        },
        "remoteTracks": {

        }
    },
    "pastIncarnations": [

    ],
    "uvv": {
        "id": 2,
        "sigHints": 0,
        "sigChain": 5,
        "cachedAt": 1498245385435
    }
}
```

**Pull request merged**